src/main/java/teammates/common/util/ActivityLogEntry.java

```diff
@@ -1,365 +1,308 @@
+
+        public Builder(String name, String url, long time) {
+            actionName = name == null ? Const.ActivityLog.UNKNOWN : name;
+            actionUrl = url == null ? Const.ActivityLog.UNKNOWN : url;
+            logTime = time;
+
+            actionResponse = actionName;
+        }
 
-        return exceptionLog.generateLogMessage();
-    }
+        public Builder withActionResponse(String val) {
+            if (val != null) {
+                actionResponse = val;
+            }
+            return this;
+        }
 
-    public boolean isTestingData() {
-        return email.endsWith(Const.ActivityLog.TEST_DATA_POSTFIX);
-    }
+        public Builder withUserRole(String val) {
+            if (val != null) {
+                userRole = val;
+            }
+            return this;
+        }
+
+        public Builder withUserName(String val) {
+            if (val != null) {
+                userName = val;
+            }
+            return this;
+        }
+
+        public Builder withUserGoogleId(String val) {
+            if (val != null) {
+                userGoogleId = val;
+            }
+            return this;
+        }
+
+        public Builder withUserEmail(String val) {
+            if (val != null) {
+                userEmail = val;
+            }
+            return this;
+        }
 
+        public Builder withMasqueradeUserRole(boolean val) {
+            isMasqueradeUserRole = val;
+            return this;
+        }
+
+        public Builder withLogId(String val) {
+            if (val != null) {
+                logId = val;
+            }
+            return this;
+        }
+
+        public Builder withLogMessage(String val) {
+            if (val != null) {
+                logMessage = val;
+            }
+            return this;
+        }
+
+        public Builder withActionTimeTaken(long val) {
+            actionTimeTaken = val;
+            return this;
+        }
+
+        public long getLogTime() {
+            return logTime;
+        }
+
+        public String getActionName() {
+            return actionName;
+        }
+
+        // use builder to build ActivityLogEntry
+        @SuppressWarnings({"PMD.AccessorMethodGeneration", "PMD.AccessorClassGeneration"})
+        public ActivityLogEntry build() {
+            ActivityLogEntry entry = new ActivityLogEntry();
+
+            entry.logTime = logTime;
+            entry.actionName = actionName;
+            entry.actionTimeTaken = actionTimeTaken;
+            entry.actionResponse = actionResponse;
+            entry.userRole = userRole;
+            entry.userName = userName;
+            entry.userGoogleId = userGoogleId;
+            entry.userEmail = userEmail;
+            entry.logMessage = logMessage;
+            entry.actionUrl = actionUrl;
+            entry.logId = logId;
+            entry.actionTimeTaken = actionTimeTaken;
+            entry.isMasqueradeUserRole = isMasqueradeUserRole;
```

Comment:
All the value are set to "Unkown" by default. The validation is done in `with*` method.